compiler/rustc_lint/src/default_could_be_derived.rs

```diff
@@ -0,0 +1,206 @@
+    /// `Default` trait of types with default field values.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![feature(default_field_values)]
+    /// struct Foo {
+    ///     x: i32 = 101,
+    ///     y: NonDefault,
+    /// }
+    ///
+    /// struct NonDefault;
+    ///
+    /// #[deny(default_overrides_default_fields)]
+    /// impl Default for Foo {
+    ///     fn default() -> Foo {
+    ///         Foo { x: 100, y: NonDefault }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Manually writing a `Default` implementation for a type that has
+    /// default field values runs the risk of diverging behavior between
+    /// `Type { .. }` and `<Type as Default>::default()`, which would be a
+    /// foot-gun for users of that type that would expect these to be
+    /// equivalent. If `Default` can't be derived due to some fields not
+    /// having a `Default` implementation, we encourage the use of `..` for
+    /// the fields that do have a default field value.
+    pub DEFAULT_OVERRIDES_DEFAULT_FIELDS,
+    Deny,
+    "detect `Default` impl that should use the type's default field values",
+    @feature_gate = default_field_values;
+}
+
+#[derive(Default)]
+pub(crate) struct DefaultCouldBeDerived {
+    data: Option<Data>,
+}
+
+struct Data {
+    type_def_id: DefId,
+    parent: DefId,
+}
+
+impl_lint_pass!(DefaultCouldBeDerived => [DEFAULT_OVERRIDES_DEFAULT_FIELDS]);
+
+impl<'tcx> LateLintPass<'tcx> for DefaultCouldBeDerived {
+    fn check_impl_item(&mut self, cx: &LateContext<'_>, impl_item: &hir::ImplItem<'_>) {
+        // Look for manual implementations of `Default`.
+        let Some(default_def_id) = cx.tcx.get_diagnostic_item(sym::Default) else { return };
+        let assoc = cx.tcx.associated_item(impl_item.owner_id);
+        let parent = assoc.container_id(cx.tcx);
+        if cx.tcx.has_attr(parent, sym::automatically_derived) {
+            // We don't care about what `#[derive(Default)]` produces in this lint.
+            return;
+        }
+        let Some(trait_ref) = cx.tcx.impl_trait_ref(parent) else { return };
+        let trait_ref = trait_ref.instantiate_identity();
+        if trait_ref.def_id != default_def_id {
+            return;
+        }
+        let ty = trait_ref.self_ty();
+        let ty::Adt(def, _) = ty.kind() else { return };
+
+        // We have a manually written definition of a `<Type as Default>::default()`. We store the
+        // necessary metadata for further analysis of its body in `check_body`.
+        self.data = Some(Data { type_def_id: def.did(), parent });
+    }
+
+    fn check_impl_item_post(&mut self, _cx: &LateContext<'_>, _impl_item: &hir::ImplItem<'_>) {
+        // Clean up.
+        self.data = None;
+    }
+
+    fn check_body(&mut self, cx: &LateContext<'tcx>, body: &hir::Body<'tcx>) {
```

Comment:
One question about this `data` based approach -- what happens when we have a nested body that is a fn? Something like:


```
impl Default for Ty {
  fn default() {
    fn other_body() -> Ty {
      Ty { /* override a default struct field or something, idk come up with a way to trigger the lint */ }
    }
  }
}
```

Might need some tweaking but I think you know what I mean.

---

If this does in fact end up triggering `check_body` for the wrong body, it may be better just to inline this logic into the `check_impl_item` function. 

It should be possible to get the hir body by just matching on the `ImplItemKind::Fn(_, body_id)` to get the id, then calling `tcx.hir.body(body_id)`.

It actually may be beneficial to do this rather than deferring to the `check_body` function just to make this code simpler.

Reply:
That is, rather than trying to split it across two functions and letting the lint infrastructure access the body for you, just access the body for yourself and make all of this live inside of one function.

Reply:
Changing that would make us go counter to https://github.com/rust-lang/rust/pull/134441#discussion_r1889314886 (use `cx.typeck_results()`) because then the `LateContext` doesn't have an `enclosed_body` yet, and we can't change it because that falls under the `LateContextAndPass` responsibility (in `rustc_lint/src/late.rs`). I can either go back to using `cx.tcx.has_typeck_results(expr.hir_id.owner.def_id)` *or* impl my own `hir::Visitor`, making the lint much more... non-standard.

Reply:
Well, you don't need to go back to using `cx.tcx.has_typeck_results(expr.hir_id.owner.def_id)`. Just call `tcx.typeck(def_id)` for the item you have.